src/core/composer/qgsaddremoveitemcommand.cpp

```diff
@@ -80,5 +80,5 @@ void QgsAddRemoveItemCommand::switchState()
     emit itemAdded( mItem );
     mState = Added;
   }
-  QgsProject::instance()->setDirty( true );
+  mComposition->project()->setDirty( true );
```

Comment:
Actually I wanted to make QgsProject association optional, but in the end scratched the idea for now. The reason is that I would need to add a bunch of extra logic to handle cases with no project and more logic when project association is changed. Also, working without project is fine various cases, but in many other cases the composition would misbehave (e.g. access to project's map layers in map item, scale bar's distance calculator config from project, legend uses project's layer tree by default). So for now I decided to always force project association with composer, and later we may decide to relax that and make project association optional... does that make sense?